### PR TITLE
kafka-util: absorb responsibility for creating topics synchronously

### LINKED ISSUE #3756: Refactor testdrive to use functionality in test_util

### PR DESCRIPTION
**Pull request by benesch**
This makes it possible to share this code between testdrive and
test-util.

Fix #3756.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/3806)

### CODE REVIEW

File: src/kafka-util/src/admin.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+/// attempting to consume from or produce to a topic immediately after its
+/// creation can result in "unknown topic" errors.
+///
+/// This function does not return successfully unless it can find the metadata
+/// for the newly-created topic in a call to [`Client::fetch_metadata`] and
+/// verify that the metadata reports the topic has the number of partitions
+/// requested in `new_topic`. Empirically, this seems to be the condition that
+/// guarantees that future attempts to consume from or produce to the topic will
+/// succeed.
+pub async fn create_topic<'a, C>(
```
Comment (ruchirK) [inline src/kafka-util/src/admin.rs:39]: so just a thought - do you think this ought to be upstreamed into rust-rdkafka?

File: src/kafka-util/src/admin.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            "kafka topic creation returned {} results, but exactly one result was expected",
+            res.len()
+        )));
+    }
+    match res.into_element() {
+        Ok(_) | Err((_, RDKafkaError::TopicAlreadyExists)) => Ok(()),
+        Err((_, e)) => Err(CreateTopicError::Kafka(KafkaError::AdminOp(e))),
+    }?;
+
+    // Topic creation is asynchronous, and if we don't wait for it to
```
Comment (ruchirK) [inline src/kafka-util/src/admin.rs:61]: I think we need to remove the bits in this comment that talk about single partition-ness because this function doesnt actually care about that afaict

File: src/kafka-util/src/admin.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    new_topic: &'a NewTopic<'a>,
+) -> Result<(), CreateTopicError>
+where
+    C: ClientContext,
+{
+    let res = client
+        .create_topics(iter::once(new_topic), &admin_opts)
+        .await?;
+    if res.len() != 1 {
+        return Err(CreateTopicError::InvariantViolation(format!(
```
Comment (ruchirK) [inline src/kafka-util/src/admin.rs:51]: as far as I can tell we only ever return `InvariantViolation` here -- the error makes sense to me if you had multiple invariants that could potentially be violated but if its just this perhaps its better as `TopicCountMismatch` or something like that? very nitpicky, very TIOLI

File: src/kafka-util/src/admin.rs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            });
+        }
+        Ok(())
+    })
+    .await
+}
+
+/// An error while creating a Kafka topic.
+#[derive(Debug)]
+pub enum CreateTopicError {
```
Comment (ruchirK) [inline src/kafka-util/src/admin.rs:93]: this is almost certainly a dumb q but what is the benefit of having a structured error enum like this instead of just `bail`-ing with strings? my understanding is that it allows the caller to more easily program different actions for different kinds of errors -- is that basically it or is there more to it?

File: src/kafka-util/src/admin.rs (hunk 1 above)
Comment (benesch) [reply src/kafka-util/src/admin.rs:39]: In principle, yes, but anything that goes in rust-rdkafka has a much higher maintenance burden because it has to work for any user of the library, not just us. I'd be willing to believe that there are still ways this topic syncing goes wrong; it's ok for us because we only need it to work in tests, but it goes into rust-rdkafka it's liable to get used in more production-like environments.

To be honest it's really silly that the create topic request can return before the topic is actually usable. There's all sorts of bugs filed against Kafka/librdkafka about this:

  * https://github.com/confluentinc/confluent-kafka-python/issues/524#issuecomment-456783176
  * https://github.com/confluentinc/confluent-kafka-python/issues/856
  * https://github.com/apache/kafka/pull/7343

tl;dr if someone else wanted to maintain this stuff in rust-rdkafka, for sure, but while it's just me I'd rather not open that can of worms.

File: src/kafka-util/src/admin.rs (hunk 3 above)
Comment (benesch) [reply src/kafka-util/src/admin.rs]: :+1: sure, done

File: src/kafka-util/src/admin.rs (hunk 4 above)
Comment (benesch) [reply src/kafka-util/src/admin.rs:89]: Yup, exactly. I did it like this purely because testdrive doesn't use anyhow. It probably should, and if it learns to then we could just bail here and things would get a lot simpler.
